### ISSUE TITLE
Cosmos DB composite Indexes

### ISSUE THREAD

**Issue by JarFrank** [Azure/azure-mobile-apps#503]
## Describe the bug

Hi!
it's a great package! I have been plaing with it and managed to create entire flow with sql server database, .net6 and xamarin. It works great!

Unfortunately, when I switched to Cosmos DB, I ran into some issues.
I cannot sync/refresh the offline items and I getting infinite loop while PullItemsAsync
1. PushItemsAsync - new items have been successfully saved
2. PullItemsAsync - infinite loop
- API try to pull items and creates odata:
```
https://my-domain.azurewebsites.net/tables/todoitemcosmos?$filter=(updatedAt gt cast(2022-11-11T20:15:25.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true
```
- this is translated to a sql query:
```
SELECT c
FROM root c
WHERE ((c["Discriminator"] = "TodoItemCosmos") AND (c["UpdatedAt"] > @__TypedProperty_0))
ORDER BY c["UpdatedAt"], c["Id"]
OFFSET 0 LIMIT @__TypedProperty_1
```
- I'm getting an error from the Cosmos DB:
```
The order by query does not have a corresponding composite index that it can be served from
```

I fixed that by creating 
```
"compositeIndexes": [
        [
            {
                "path": "/UpdatedAt",
                "order": "ascending"
            },
            {
                "path": "/Id",
                "order": "ascending"
            }
        ]
    ]
```
The question is, do we need this additional order by id?

## To Reproduce

1. Create projects based on the samples and https://learn.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#azure-cosmos-db

## What platforms?

* Env:
  * ASP.NET CORE 6
  * DataSync 5.12
  * Database: Cosmos DB

* Client (.NET, JavaScript, iOS, Android):
  * Xamarin.Forms
  * DataSync 5.12
  * Tested only on the emulator

**Comment by adrianhall** [comment]
Can you upload your project to GitHub and give me access so I can take a look?

**Comment by adrianhall** [comment]
After trying this out, I could not re-produce the situation.  I did add my bicep infrastructure files to the CosmosTodoService sample, in case it was that your database wasn't set up correctly.  Without taking a look at the code, I cannot comment further.

If you do upload the code and give me access, re-open this issue with the information.  I can then further diagnose the problem.

**Comment by Joost1982** [comment]
Today I reproduced the error(s). Here are the steps I took:

- Clone project from Github

1.
- Run bicep to create CosmosDb + container with name 'AppDbContext' and partitionkey '/id'
- Post an item with Postman gives this error: _PartitionKey extracted from document doesn't match the one specified in the header._
- Get items with '?$orderby=updatedAt&$count=true_' gives this error: _The order by query does not have a corresponding composite index that it can be served from._

2.
- Remove container and just start the CosmosTodoService app; this will create the container now with the partitionkey '/__partitionKey'
- Post item with Postman now works
- Get items with '?$orderby=updatedAt&$count=true' gives the same error: 'The order by query does not have a corresponding composite index that it can be served from.'
- After adding the composite index like @JarFrank suggested, getting items work.

**Comment by omkelderman** [comment]
to add to what @Joost1982 said: the fact that it returns that erroneous response results in the infinite loop when calling PullItemsAsync as mentioned at the top of this page. That result object is weird since the items array is empty when it shouldnt be and the nextLink contains the exact link as the original request, hence the infinite loop.

This is imo a bigger problem than the fact that composite index is needed, and the actual bug: The fact that cosmos errors out but the controller does not produce an error. I would have expected something like a 500 response, not a response that contains no data when it should have had data.

If you're not watching the server app console for errors, it is quite frustrating debugging this from the client side of things since that response makes no sense and at first I didnt understand why it was doing that.

**Comment by adrianhall** [comment]
I'll take another look.

Can someone post the exact versions of libraries they are using?

**Comment by omkelderman** [comment]
I'll whip up a sample project with repro instructions that describes the problem as I observe it and upload it to github once im done with work

**Comment by omkelderman** [comment]
@adrianhall https://github.com/omkelderman/AzureMobileAppsIssue503 lemmy know if you have any questions.

I did a bit of a writeup with how to run the project to get the same results as I have, with some commentary mixed in about what I expected the result to be and where it deviates from my expectations and what I considered to be the bug

**Comment by adrianhall** [comment]
Fix is released in v5.1.7.

Docs are updated and will be included in tomorrow mornings publication run.
